sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OtlpConfigUtil.java

```diff
@@ -46,6 +50,17 @@ static void configureOtlpExporterBuilder(
     String endpoint = config.getString("otel.exporter.otlp." + dataType + ".endpoint");
     if (endpoint == null) {
       endpoint = config.getString("otel.exporter.otlp.endpoint");
+      String protocol = getOtlpProtocol(dataType, config);
+      if (endpoint != null && protocol.equals(PROTOCOL_HTTP_PROTOBUF)) {
+        URI endpointUri;
+        try {
+          endpointUri = new URI(endpoint);
+        } catch (URISyntaxException e) {
+          throw new ConfigurationException("Invalid endpoint, must be a URI: " + endpoint, e);
+        }
+        String signalPath = signalPath(dataType);
+        endpoint = endpointUri.resolve(signalPath).toString();
```

Comment:
Won't `resolve` with a leading slash in the argument strip of the entire path part of the URL, leaving only the hostname? The tests don't cover that. I suspect `URI.create("http://localhost:4317/collector").resolve("/v1/traces")` will result in http://localhost:4317/v1/traces`, *not* the probably expected `http://localhost:4317/collector/v1/traces`.

Reply:
Yeah why I asked for the test :-)

> I think that would violate the spec. Configuring a per-signal endpoint should not append anything, so http://localhost:4317/collector should stay as just that.

I agree with this and think the spec allows for it. I guess it means we get something like this

`http://localhost:4317` -> `http://localhost:4317/v1/traces`
`http://localhost:4317/` -> `http://localhost:4317/` (otherwise, there'd be no way to host at a root like `http://otlp-http.com/`, not appending if a root path is passed explicitly seems like the only way to support it)
`http://localhost:4317/collector` -> `http://localhost:4317/collector`

Reply:
No, the difference is not the slash but whether you set the shared configuration for all signals (auto-appending) or the per-signal config (URL is used as-is).

Reply:
Hmm, it *is* relatively clear on that part:

> When using OTEL_EXPORTER_OTLP_ENDPOINT, exporters SHOULD follow the collector convention of appending the version and signal to the path

Note the
> When using OTEL_EXPORTER_OTLP_ENDPOINT

This exludes the OTEL_EXPORTER_OTLP_signal_ENDPOINT vars

https://github.com/open-telemetry/opentelemetry-specification/blob/v1.6.1/specification/protocol/exporter.md#configuration-options

Reply:
Ah ok that sounds pretty good then

Reply:
I had this test in my initial commit, which I've reverted to (along with using string concatenation instead of URI.resolve().

The weakness of string concatenation is that the behavior will be wrong if there is a query string or a fragment included. The spec appears to disallow query strings / fragments with this line:

> The endpoint MUST be a valid URL with scheme (http or https) and host, and MAY contain a port and path.

I.e. by not explicitly allowing a query string / fragment the assumption is that they're not allowed. The best logic would:
1. parse the input as a `URL`
2. confirm there aren't any parts included that aren't explicitly allowed
3. get the path from the URL, and manipulate it to include include the signal type path (accounting for trailing slashes)
4. rebuild a new url from the parts, using the newly constructed path, and convert it to a string

I can do this. Let me know if ya'll are interested or if you think its overkill.

Reply:
That seems like a good plan. And, having unit tests to cover all the possible strange cases will be very useful documentation, at least internally.